Add ERROR message and first Bus messages

bertrand-marquis opened the pull request:
This pull requests contains 3 patches:
- introduce a bus chapter: we need to decide how and where this should actually be done
- create the virtio-msg ERROR message
- first attempt to create Bus messages

I added some TODOs as there are various subjects that we will need to discuss there.

bertrand-marquis committed:
Define the message bus

DRAFT

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Introduce virtio-msg ERROR message

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
First attempt for bus content

Add bus messages definitions and several TODO to be discussed.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Remove Connect and Disconnect

Remove Connect and Disconnect messages.
Re-work the messages numbering:
- Error is now 0x1
- Mark 0x0 as Reserved
- Move messages down to use all numbers
- Move Event messages from 0x10 to 0x20 to leave more space for future
  non even messages.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Add a more information in Error message

Define a Message ID field in error message to carry the ID of the
message that triggered an error.
Define a Data Type field in Error message and a value STRING (1) so that
the rest of the payload of the message can be used as NULL terminating
string that can be printed.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Instruct to use ERROR message in FEATURE_GET

When the driver is requesting feature at an index at which no features
are available and after which no more features are available, instruct
to return an ERROR message so the driver knows that it does not need to
scan further.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Cleanup bus message introduction

Make the chapter and content a bit more coherent with the main
virtio-msg chapter.
While there also reserve message ID 0 and move ERROR message to 1.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Rework the Bus error message

Make the bus error message more coherent with the virtio-msg one and add
a possible string payload and a payload type paramter.

Add some clarification on when and by who this can be sent.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Rework Get devices Bus message

Make clear that the offset is not a multiple of 32 (ie offset 1 gives
availibility of device 32 to 63).

Fix some wording and latex issues.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Add message definition for device added and removed

Add content and message definition for bus messages signaling that a
device has been added or removed.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Add bus ping message definition

Add ping message description and definition with 32bit of data that must
be copied back in the answer.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Define the bus status message

Add a definition of the bus status message (replaces reset message).

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis commented:
I added more patches to handle the changes discussed during the HVAC meeting on the 30/01/2025.
Please give me comments :-)

arnopo commented on transport-msg.tex:
```diff
@@ -90,8 +90,15 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+extra printable information.
 
 This message can be received by the driver or the device and can be sent as an
 answer to an other message by the driver or the device or any element between
```
I would move this above  "The message contain an error code defined using the Standard Posix...."

bertrand-marquis replied:
Actually the type is coming after the error code in the message content so moving this above might weird.
@arnopo Any reason why you want to move this up before the error code ?

arnopo replied:
I was speaking only about this sentence:
```
This message can be received by the driver or the device and can be sent as an
answer to an other message by the driver or the device or any element between
```
It seems to me that this sentence describes the use of the error message, while the two previous ones speak about the message content."
that's why I propose tu movz it before

bertrand-marquis replied:
Ok now i understand. You want me to move this sentence before the parameters description.
Make sense, I will do that.

arnopo commented on transport-msg.tex:
```diff
@@ -90,8 +90,15 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
 The Error message is used to signal to the driver or the device that a message
-it wanted to send could not be sent. The message contain an error code defined
-using the Standard Posix Error Codes (TODO: add a reference).
+it wanted to send could not be sent.
```
The sentence 'a message it wanted to send could not be sent' seems to me not correct vs the usage
 proposal
"The error message is used to signal to the driver or the device that a received message is not valid or cannot be addressed. It can also be used to report an internal error to the other entity."

bertrand-marquis replied:
ack

arnopo commented on transport-msg.tex:
```diff
@@ -448,26 +449,26 @@ \subsection{Message Bus}\label{sec:Virtio Transport Options / Virtio Over Messag
-devices.
+This can be done in a static way, using a device tree for example, or in a way
+using messages to discover from the frontend side which device IDs are
+available on the backend side (using the BUS_MSG_GET_DEVICES bus message).
```
proposal:
```
  - The frontend(driver) may use a predefined virtio devices ID list ( static way), or may use BUS_MSG_GET_DEVICES  message to discover  devices IDs available on the device (dynamic way.
  - The backend(device)  may use a predefined virtio devicesID list that can be provided to the driver on BUS_MSG_GET_DEVICES  request , or may announce deviceIDs using   BUS_MSG_DEVICE_ADDED  and BUS_MSG_DEVICE_REMOVED messages.
```

jamesr replied:
I think we want to allow defining buses statically and dynamically so I'd phrase it as an "and" rather than a "or"

bertrand-marquis replied:
ack and ack

arnopo commented on transport-msg.tex:
```diff
@@ -437,3 +429,249 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+The Get Devices message is used by the frontend to retrieve the ID of the
+devices available on the backend of the bus.
+
+The frontend requests the availibility at an offset and the backend returns a
```
should we use  "word selection" instead of "offset" term as done for the virtio MMIO? ( see GuestFeaturesSel as example)

bertrand-marquis replied:
I will rephrase

bertrand-marquis replied:
@arnopo word selection in this context looks a bit weird. The FeatureSel is actually selecting a "Word" (ie 32bis) using a different register (ie you write in feature sel to configure where you read using deviceFeatures) where here we have send a message to request a list of devices availability.
I am thinking of changing the sentence this way:
The frontend requests the availibility of a group of 256 device IDs and the
backend returns a bitmap indicating which device IDs are available at the
requested device ID offset.
Tell me what you think.

arnopo replied:
I'm okay with your proposal. I added two new comments on this message description.

arnopo commented on transport-msg.tex:
```diff
@@ -437,3 +429,249 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+This message shall be used in the following way:
+\begin{itemize}
+\item frontend request devices at offset 0.
+\item backend answers with the availibility of devices 0 to 31 and indicates to
```
\item frontend request devices at offset x.
\item backend answers with the availability of devices x*32  to x*32 +31 and indicates to

bertrand-marquis replied:
yes make sense to do a more generic example. I will modify this.

arnopo commented on transport-msg.tex:
```diff
@@ -437,3 +429,249 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+
+A request at offset X give the availability of devices X*32 to X*32 + 31.
+
+The backend also return a next offset that indicates to the frontend what
```
typo   s/give/gives s/return/returns/

bertrand-marquis replied:
ack

arnopo reviewed (commented):
Hi @bertrand-marquis
few comments on the form only else LGTM
Thanks!

jamesr commented on transport-msg.tex:
```diff
@@ -150,9 +142,9 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
- index ? is it ok to just specify that 0 will be returned in that case ? can
- we expect the driver to always know how much features a device has ?
+If there are no features available at and after the requested index, the
+ERROR message can be used by the device to signal to the driver that there is
```
As a general note, I recommend avoiding the passive voice whenever possible to make it clearer what part of the system is responsible for taking a specific action. In this case, I can interpret the requirement two ways. If the device is permitted to either send an all-zero response or an error I would reword this to say:

If there are no features available at the requested index or after the requested index the device SHOULD send an ERROR message to the driver to signal that there is no need to scan for features at or above the requested index.

If it the device is required to send this message change SHOULD to a MUST.

bertrand-marquis replied:
Agree i will avoid the passive voice.
On a second thought on this specific example, I am not sure it is really a good idea to give several solutions here so i will remove the error possibility and stick to an all zero response here.

jamesr commented on transport-msg.tex:
```diff
@@ -437,3 +429,249 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+The Message Bus transmits the messages between the frontend and the backend in
+both directions.
+
+A system can have several Message Busses and it shall respect the following
```
"buses" is a much more common spelling than "busses" both in virtio and in general English use.

bertrand-marquis replied:
ack

jamesr commented on transport-msg.tex:
```diff
@@ -437,3 +429,249 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+
+The Error message is used by one side of the bus to inform the other side that
+one message could not be processed correctly. The message contains an error
+code defined using the Standard Posix Error Codes (TODO: add a reference).
```
POSIX does not define the numeric value of the errors that it defines (other than reserving zero):

https://pubs.opengroup.org/onlinepubs/9799919799/functions/V2_chap02.html#tag_16_03

Specific implementations pick the numeric values used by different errors on their platform and it's quite messy in practice. Since this protocol will be used to communicate between systems that are running different platforms they may not use the same numeric values for errors. This specification needs to define the actual numeric values for errors that are valid in this protocol.

bertrand-marquis replied:
That is a very good point.
We need to come up with a list of errors and define them in the spec instead of the current system pointing to POSIX.
I do not think we need the full POSIX list of errors so we need to come with a reduce subset, at the end there is always the string or IMPDEF solution if an implementation wants to provide more information.
Anybody has something we could start from as a list of error codes ?

jamesr replied:
I noticed that another part of the draft (from admin.tex) defines a set of numeric error codes and says that the values were picked to match what Linux uses for similar errors:

https://github.com/Linaro/virtio-msg-spec/blob/c992db3e403989ae7253b6e2312339f193ccbdef/admin.tex#L174

Perhaps that's a model we could follow? I don't know where that text is from or where it was discussed. It'd be good to have a generic for errors across virtio.

bertrand-marquis replied:
Yes it would be a good idea to have some Virtio error code defined in a generic chapter and have admin and virtio message refer to it.
@stsquad Any advice here on how we should propose that ?
We could make an independent patch creating virtio Errors and removing admin ones to refer to it so that we could also refer to it in our patches. Should it be part of our RFC serie ?

bertrand-marquis committed:
Handle review comments

Handle review comments.
I did that in a new patch to go quicker as we will squash everything at
the end anyway.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Define a list of error codes

Introduce a list of error codes instead of referring to POSIX ones and
use them in both ERROR messages (transport and bus).

The list of errors must be discussed and some might need to be added or
removed.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

arnopo commented on transport-msg.tex:
```diff
@@ -437,3 +455,256 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+Answer & 0 & 2 & Offset \\
+       & 2 & 2 & Next offset or 0 if no devices after Offset \\
+       & 4 & 2 & Reserved (MBZ) \\
+       & 6 & 32 & Bit[n]: Device[offset*32 + n] not available(0)/available(1) \\
```
the max size of the payload is 36.   except if i missed something:

```diff
Answer & 0 & 2 & Offset \\
       & 2 & 2 & Next offset or 0 if no devices after Offset \\
-       & 4 & 2 & Reserved (MBZ) \\
-       & 6 & 32 & Bit[n]: Device[offset*32 + n] not available(0)/available(1) \\
+       & 4 & 32 & Bit[n]: Device[offset*32 + n] not available(0)/available(1) \\
```

bertrand-marquis replied:
Right i need to make a pass on all messages.
Thanks

bertrand-marquis replied:
Actually a Bus message has 38 bytes of payload, a Virtio Message only has 36 as we have the Device ID added in the header.

arnopo replied:
Do you mean that the header would be dependent on VIRTIO_MSG_TYPE? It seems to me that this would complicate the description. 
Do you need 38 bytes of payload for a message?

If there is no specific reason, I would be in favor of keeping the same header for both types and updating the header description.


```diff
\begin{tabularx}{\textwidth}{|l|l|l|X|}
\hline
Name & Offset & Size (bytes) & Content \\
\hline \hline
VIRTIO_MSG_TYPE & 0 & 1 & Bit[0]: 0=Request, 1=Answer \newline Bit[1]: 0=virtio-msg, 1=bus-msg \newline Bit[2-7]: Reserved \\
\hline
VIRTIO_MSG_ID & 1 & 1 & Message ID \\
\hline
- VIRTIO_MSG_DEV_ID & 2 & 2 & Device ID \\
+ VIRTIO_MSG_DEV_ID & 2 & 2 & Device ID, must be set to 0 if VIRTIO_MSG_TYPE Bit[1]=1
\hline
VIRTIO_MSG_PAYLOAD & 4 & 36 & Payload \\
\hline
\end{tabularx}

```

bertrand-marquis replied:
We have no message requiring 38 bytes so I do not have anything against that at all.
I will do the changes for that.

arnopo commented on transport-msg.tex:
```diff
@@ -437,3 +455,256 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+backend returns a bitmap indicating which device IDs are available at the
+requested device ID offset.
+
+A request at offset X gives the availability of device IDs X*32 to (X+1)*32-1.
```
It is quite confusing between the offset in bytes and the device ID that is a bitmap. If we return a 256-bit bitmap in 32 bytes, it seems to me that the equation should be: 
`device IDs X*256 to (X+1)*256-1.`

bertrand-marquis replied:
Yes this is wrong in fact, I will fix that

bertrand-marquis committed:
Various fixes

- Move sentence earlier in error message
- Fix some inconsistencies in message name formatting
- Fix bus message content description to fit the 38 Bytes payload
- Fix GET_DEVICES message to use 256 devices per message and not 32

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis commented:
I pushed a last change to handle the comments.
I plan to squash all patches together after as we will submit them as one patch. This will also make review easier for everyone.

bertrand-marquis committed:
Reuse the message format for Bus messages

Define that bus messages must have the Device ID field set to 0 and
reuse the same format than the virtio messages.

With that change reduce the payload size used by Bus messages to 36
bytes and just reuse the same format for the Error message.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
